### PR TITLE
WIP: fix #558 by migrating w04 exercise to Scala 3

### LINKED ISSUE #558: Migrate exercise w04 objects to Scala 3

### PR DESCRIPTION
**Pull request by snctfd**
Still have to test some parts in WSL 2 (WSL 1 doesn't play nice with AWT) and double check that I haven't missed anything, then this should be ready for merge.

### CODE REVIEW

**Comment by bjornregnell** [comment]
ok thanks; let me know when it's ready for review; don't forget to make a complete sbt build to check everything...

**Comment by snctfd** [comment]
Was not able to get an X server up and running but I did spot a few bits of Scala 2 syntax that I had initially missed. As of now, the bits that require PixelWindow have not been tested, but the chapter should be at least good enough for the test print.

File: compendium/modules/w04-objects-exercise.tex [hunk 1]
```diff
@@ -332,7 +319,7 @@
 
 \begin{REPL}
 > wget -O introprog_2.13-1.1.4.jar http://cs.lth.se/pgk/lib
```
Comment (bjornregnell) [inline compendium/modules/w04-objects-exercise.tex:321]: There is new jar for Scala 3 introprog_3-1.1.5.jar, but that needs to be uploaded to http://cs.lth.se/pgk/lib 
Issue here: https://github.com/lunduniversity/introprog-scalalib/issues/22

File: compendium/modules/w04-objects-exercise.tex [hunk 2]
```diff
@@ -485,17 +471,15 @@
 \end{minipage}
 
 \begin{REPL}
-> scala -cp introprog_2.13-1.1.4.jar
+> scala -cp introprog_2.13-1.1.4.jar -repl
```
Comment (bjornregnell) [inline compendium/modules/w04-objects-exercise.tex:474]: should be updated eventually; issue here: https://github.com/lunduniversity/introprog/issues/561

File: compendium/modules/w04-objects-exercise.tex [hunk 3]
```diff
@@ -485,17 +471,15 @@
...
 scala> def rak(p: Pt)(d: Int) = w.line(p._1, p._2, ???, ???, color)
-scala> def fyll(p: Pt)(s: Int) = for (i <- ???){ rak((p._1, ???))(s) }
+scala> def fyll(p: Pt)(s: Int) = for i <- ??? do rak((p._1, ???))(s)
 
-scala> :paste
-object Color {
- ???
-}
+scala> object Color:
+     |   ???
```
Comment (bjornregnell) [inline compendium/modules/w04-objects-exercise.tex:482]: I have removed all `|` in prints as they dont add anything and just make copy paste harder

File: compendium/modules/w04-objects-exercise.tex [hunk 4]
```diff
@@ -822,36 +795,26 @@
...
-object Main {
-  def main(args: Array[String]): Unit = {
-    import p1._
+object Main:
+  def main(args: Array[String]): Unit =
+    import p1.*
     p11.hello.hello
     p12.hello.hello
-    import p2.{p21 => apelsin}
+    import p2.{p21 as apelsin}
```
Comment (bjornregnell) [inline compendium/modules/w04-objects-exercise.tex:816]: with only one rename no need for braces

File: compendium/modules/w04-objects-exercise.tex [hunk 5]
```diff
@@ -1113,7 +1078,7 @@
 
 \Subtask Använd funktionen \code{introprog.Dialog.select} för att be användaren välja mellan sten, sax och påse. Vad är returtypen?
 % \begin{REPL}
-% scala> import JOptionPane.{showOptionDialog => showOpt}
+% scala> import JOptionPane.{showOptionDialog as showOpt}
```
Comment (bjornregnell) [inline compendium/modules/w04-objects-exercise.tex:1081]: no braces needed

**Comment by bjornregnell** [comment]
I have some comments, but that can be fixed in another PR. I'll merge this for now. Please keep track of my comments for a later PR.

**Comment by bjornregnell** [comment]
also good if you can test introprog.PixelWindow in "pure" windows 10 without wsl2 using a coursier install and some Windows-adapted workaround of https://github.com/coursier/coursier/issues/2058#issuecomment-866792591  perhaps inspired by some line from the black magic bat-stuff here https://github.com/michelou/dotty-examples

**Pull request merged**